Fix gutterless, closes #134

triblondon opened the pull request:
![image](https://cloud.githubusercontent.com/assets/1735391/15564109/7fb8aa32-2349-11e6-86a8-9950d269d79d.png)

![image](https://cloud.githubusercontent.com/assets/1735391/15564114/89e28500-2349-11e6-87f6-e3067d6522ce.png)

kaelig commented:
This unfortunately breaks sub grids.

My recommendation would be to have a modifier on containers, removing left and right padding.

Something like this:

```scss
	.o-grid-container--bleed {
		padding-left: 0;
		padding-right: 0;
	}
```

triblondon commented:
@kaelig can you elaborate a bit on how it breaks subgrids?

OK, so the bleed idea would leave us with a solution where:

* There is a default padding on container...
* removed by the modifier...
* then the row has negative margin which pushes it outside the container...
* and the column has padding that brings it back to 0 again

I'd be concerned about possible overflow scrollbars coming in, unless there was overflow hidden on the container.  But if that's the only solution that works for all of us and it doesn't cause any side effects, fair enough.

kaelig commented:
The body should have overflow-x: hidden on it. It's pretty common to do it on responsive sites to make sure no third party content gets out of the desired viewport and adds horizontal scrollable areas. 

Subgrids are rows inside of cells, there are a few on the test page, but not for compact rows (yet). With the technique from this PR, I found that subgrids got out of their container. If that makes no sense I can throw a bit of code your way.

triblondon commented:
OK, updated to use new approach.  Also updated the demo to remove margin highlighting because it is misleading now that it goes outside the container and is in some cases negative.

triblondon commented:
Still needs documenting, if Origami team are OK with this.

kaelig commented on src/scss/_main.scss:
```diff
@@ -464,6 +464,11 @@
 		}
 	}
+
+	&--bleed {
```
Careful with this notation, it makes files hard to grep

onishiweb replied:
Because of the way the mixin is being used here and the class it would need to generate I think this is ok

kaelig replied:
Fair point!

kaelig replied:
But how do I use this in Sass? Shouldn't the bleed should be a parameter, and this class be generated separately as part of the "generate" mixin?

onishiweb replied:
Good point, quite possibly. I'll have a look into that.

triblondon replied:
OK, I changed it to allow use as a mixin

kaelig commented on demos/src/scss/_demos.scss:
```diff
@@ -48,10 +48,9 @@ body:before {
 .o-grid-container {
-	background: lighten(#f6e9d8, 8.5%);
+	background: #EAE8DF;
 }
```
These backgrounds have proven super useful when doing visual tests, maybe it'd be best to make a special case for the bleed modifier that removes the background on the sides and the direct children rows?

onishiweb replied:
No major opinion here, I think both look ok. The row colouring would make debugging easier, but can easily be added when doing so, @triblondon any reason for changing them?

triblondon replied:
The current architecture means that the demo is non-symmetrical, which looks like a bug:

![image](https://cloud.githubusercontent.com/assets/1735391/15596532/36d39c7c-2403-11e6-881c-6da1ba88ae22.png)

Note that the row appears to touch the grid container on the left, but not on the right.  This is in fact the way the grid is currently designed, so isn't a bug, but looks like one.  I thought it was therefore misleading to colour the row backgrounds, which is why I removed that below.  Changing the one above was more just because on my screen it is too light to be discernable, so I can't see it.

wheresrhys replied:
Has anyone ever said it looks like a bug? I think it's useful to show these 'seams' - aids comprehension and, as kaelig says,  a useful debugging tool

onishiweb replied:
I think I agree with @triblondon's rational, while those that understand this method for doing grids, others might not and it might look like a bug. With the goal of getting people started with components more easily I think this works. Those that need to debug can easily add a row colour through the inspector or editing the SCSS themselves.

kaelig replied:
Your call, but having spent weeks working on this grid (and Rhys as well), I'd like to say that this is the kind of detail that makes the debugging experience much better and spotting the tiniest mis-alignment much easier.

onishiweb replied:
I understand your point, but this demo isn't for people that are working on the component. It's for users of the component, and our aim over the coming months will be to make using components a lot easier. Those that are working on the component can add a colour to the row for working/debugging should they need to quite easily.

kaelig replied:
Nice! Looking forward to seeing the results.

AlbertoElias replied:
How about a documented `$o-grid-debug` variable set manually to false in the demos, that if true activates the gutter colouring?

triblondon replied:
I would say that if debugging, you have devtools to highlight things as you need to. I agree with Adam, the demo is for consumers.  Adding that variable is more complexity that I don't think is necessary.

kaelig commented on demos/src/layout.mustache:
```diff
@@ -219,3 +219,16 @@
+<p style='text-align:center'>With container bleed:</p>
+
+<div class="o-grid-container o-grid-container--bleed">
+	<div class="o-grid-row o-grid-row">
```
This has twice the same class?

kaelig commented:
Ideally there would be a new test for it too! I believe @wheresrhys knows well how they work.

wheresrhys commented:
I'm happy to help with explaining the tests, but no time to write one I'm afraid

triblondon commented:
OK, the test framework was pretty broken, actually.  For example:

* a centered column was failing a margin test because the expected margin was NaN.  That was because we were trying to multiply a number by a gutter which was coming back as '20px'.  I added a parseInt to that.
* It passed, but only because the `almostEqual` function is allowing a 100% variance in the value, so e.g. if expected is '300', an acceptable value was between 0 and 600; I reduced the variance allowed to 1% anywhere that `almostEqual` is used.
* Now that test fails again, but it seems the test is wrongly including the gutter in the expect, since the outer gutter is padding on the container, not on the row. So I removed the gutter from the calculation of the expected value.
* Now it passes again, this time I think correctly.

Also added documentation for bleed.

kaelig commented:
Good stuff, thanks for doing this, it's good to see the grid evolve with sensible use cases!
